src/runtime-rs/crates/resource/src/share_fs/share_virtio_fs_standalone.rs

```diff
@@ -0,0 +1,197 @@
+            let mut extra_args: Vec<String> = self
+                .config
+                .virtio_fs_extra_args
+                .iter()
+                .map(String::from)
+                .collect();
+            args.append(&mut extra_args);
+        }
+
+        Ok(args)
+    }
+
+    async fn setup_virtiofsd(&self) -> Result<()> {
+        let sock_path = generate_sock_path(&self.config.jail_root);
+        let args = self.virtiofsd_args(&sock_path).context("virtiofsd args")?;
+
+        let mut cmd = Command::new(&self.config.virtio_fs_daemon);
+        let child_cmd = cmd.args(&args).stderr(Stdio::piped());
+        let child = child_cmd.spawn().context("spawn virtiofsd")?;
+
+        // update virtiofsd pid{
+        {
+            let mut inner = self.inner.write().await;
+            inner.pid = child.id();
+        }
+
+        let (tx, mut rx): (Sender<Result<()>>, Receiver<Result<()>>) = channel(100);
+        tokio::spawn(run_virtiofsd(child, tx));
+
+        // TODO: support timeout
+        match rx.recv().await.unwrap() {
+            Ok(_) => {
+                info!(sl!(), "start virtiofsd successfully");
+                Ok(())
+            }
+            Err(e) => {
+                error!(sl!(), "failed to start virtiofsd {}", e);
+                self.shutdown_virtiofsd()
+                    .await
+                    .context("shutdown_virtiofsd")?;
+                Err(anyhow!("failed to start virtiofsd"))
+            }
+        }
+    }
+
+    async fn shutdown_virtiofsd(&self) -> Result<()> {
+        let mut inner = self.inner.write().await;
+
+        if let Some(pid) = inner.pid {
+            let pid = ::nix::unistd::Pid::from_raw(pid as i32);
+            if let Err(err) = ::nix::sys::signal::kill(pid, nix::sys::signal::SIGKILL) {
+                if err != ::nix::Error::Sys(nix::errno::Errno::ESRCH) {
+                    return Err(anyhow!("failed to kill virtiofsd pid {} {}", pid, err));
+                }
+            }
+        }
+        inner.pid = None;
+
+        Ok(())
+    }
+}
+
+async fn run_virtiofsd(mut child: Child, tx: Sender<Result<()>>) -> Result<()> {
+    let stderr = child.stderr.as_mut().unwrap();
+    let mut stderr_reader = BufReader::new(stderr);
+
+    loop {
+        let mut buffer: String = String::new();
+        match stderr_reader.read_line(&mut buffer).await {
+            Ok(0) => {
+                info!(sl!(), "virtiofsd process exited");
+                break;
+            }
+            Ok(_len) => {
+                let trim_buffer = buffer.trim_end();
+                if !trim_buffer.is_empty() {
+                    info!(sl!(), "source: virtiofsd {}", trim_buffer);
+                }
+                if buffer.contains("Waiting for vhost-user socket connection") {
+                    tx.send(Ok(())).await.unwrap();
+                }
+            }
+            Err(e) => {
+                tx.send(Err(anyhow!("read virtiofsd console failed with: {}", e)))
```

Comment:
My understanding is, In `run_virtiofsd` the sender `tx` is used twice(if an error occurred), but the peer receiver in the caller should have been dropped after some time that virtiofsd started(only one `recv` on the receiver side in the caller), so unwrapping may lead to a panic.

There may be need a task to watch if the virtiofsd process stopped unexpectedly, refers to the current go version https://github.com/kata-containers/kata-containers/blob/6837ab72137bb627d2f348348c60a9e634aa78fc/src/runtime/virtcontainers/virtiofsd.go#L149-L160

Suggest adding a `todo` somewhere and implementing it later.

Reply:
Thanks for your review, raise an issue:
https://github.com/kata-containers/kata-containers/issues/4004